Update to C# Quickstart-Chat Server Module and Client SDK tutorial documents (#170)
clockworklabs/spacetime-docs: docs/sdks/c-sharp/quickstart.md

This second case means that, even though the module only ever inserts online users, the client's `User.OnInsert` callbacks may be invoked with users who are offline. We'll only notify about online users.

> `OnInsert` and `OnDelete` callbacks take two arguments: the altered row, and a `EventContext`. This will be `null` for rows inserted when initializing the cache for a subscription. The `EventContext.Event` is an enum autogenerated by `spacetime generate` with a variant for each reducer defined by the module. For now, we can ignore this argument.

Reviewer: ```suggestion
`OnInsert` and `OnDelete` callbacks take two arguments: an `EventContext` and the altered row. The `EventContext.Event` is an enum which describes the event that caused the row to be inserted or deleted. All SpacetimeDB callbacks accept a context argument, which you can use in place of your top-level `DbConnection`.
```

Whenever we want to print a user, if they have set a name, we'll use that. If they haven't set a name, we'll instead print the first 8 bytes of their identity, encoded as hexadecimal. We'll define a function `UserNameOrIdentity` to handle this.

```csharp

> string UserNameOrIdentity(User user) => user.Name ?? user.Identity.ToString();//[..8];

Reviewer: ?

Author: This was a debug change that I missed. Corrected!

Each reducer callback takes one fixed argument:

> The `ReducerEventContext` of the callback which contains an `Event` that contains several fields. The ones we care about are:

Reviewer: ```suggestion
The `ReducerEventContext` of the callback, which contains an `Event` that contains several fields. The ones we care about are:
```

> 1. The `CallerIdentity` is the `Identity` of the client that called the reducer.

Reviewer: ```suggestion
1. The `CallerIdentity`, the `Identity` of the client that called the reducer.
```

2. The `Status` of the reducer run, one of `Committed`, `Failed` or `OutOfEnergy`.

> 3. The error message, if any, that the reducer returned.

Reviewer: This has been changed so that the error message is now nested inside of `Status.Failed`. Could you rephrase?

Author: Corrected

```csharp
void OnConnected(DbConnection conn, Identity identity, string authToken)
{
    local_identity = identity;
    AuthToken.SaveToken(authToken);

    conn.SubscriptionBuilder()
        .OnApplied(OnSubscriptionApplied)
        .SubscribeToAllTables();
}
```

You can also subscribe to specific tables using SQL syntax, e.g. `SELECT * FROM my_table`. Our [SQL documentation](/docs/sql) enumerates the operations that are accepted in our SQL syntax.

## Connect Error callback

Should we get an error during connection, we'll be given an `Exception` which contains the details about the exception. To keep things simple, we'll just write the exception to the console.

```csharp
void OnConnectError(Exception e)
{
    Console.Write($"Error while connecting: {e}");
}
```

## Disconnect callback

> When Disconnecting, the callback contains the connection details and if an error occurs, it will also contain an `Exception`. If we get an error, we'll write the error to the console, if not, we'll just write that we disconnected.

Reviewer: ```suggestion
When disconnecting, the callback contains the connection details and if an error occurs, it will also contain an `Exception`. If we get an error, we'll write the error to the console, if not, we'll just write that we disconnected.
```

Either it's a symbol, in which case it gets code-formatted, or it's an English word, in which case it doesn't get capitalized.

1. Connect to the module. We'll store the SpacetimeDB host name and our module name in constants `HOST` and `DB_NAME`. We will also store if SSL is enabled in a constant called `SSL_ENABLED`. This only needs to be `true` if we are using `SpacetimeDB Cloud`. Replace `<module-name>` with the name you chose when publishing your module during the module quickstart.

`Connect` takes an auth token, which is `null` for a new connection, or a stored string for a returning user. We are going to use the optional AuthToken module which uses local storage to store the auth token. If you want to use your own way to associate an auth token with a user, you can pass in your own auth token here.
1. Loop until the thread is signaled to exit, calling `Update` on the SpacetimeDBClient to process any updates received from the module, and `ProcessCommand` to process any commands received from the input loop.

2. Loop until the thread is signaled to exit, calling `Update` on the SpacetimeDBClient to process any updates received from the module, and `ProcessCommand` to process any commands received from the input loop.

> 3. Finally, Close the connection to the module.

Reviewer: Now that there's only two steps here, one of which is "close," this numbered list structure is pretty silly. Can you rewrite to just say what it does, and then "Afterwards, close the connection," or similar?

Author: Corrected with better phrasing

> 2. Finally, Close the connection to the module.

Reviewer: ```suggestion
2. Finally, close the connection to the module.
```

## What's next?

Congratulations! You've built a simple chat app using SpacetimeDB.

> You can find the full code for this client [in the C# SDK's examples](https://github.com/clockworklabs/com.clockworklabs.spacetimedbsdk/tree/master/examples~/quickstart/client).

Reviewer: ```suggestion
You can find the full code for this client [in the C# client SDK's examples](https://github.com/clockworklabs/com.clockworklabs.spacetimedbsdk/tree/master/examples~/quickstart/client).
```

> Check out the [C# SDK Reference](/docs/sdks/c-sharp) for a more comprehensive view of the SpacetimeDB Rust SDK.

Reviewer: ```suggestion
Check out the [C# client SDK Reference](/docs/sdks/c-sharp) for a more comprehensive view of the SpacetimeDB C# client SDK.
```
